apps/opencs/view/render/worldspacewidget.cpp

```diff
@@ -320,49 +318,72 @@ void CSVRender::WorldspaceWidget::addEditModeSelectorButtons (CSVWidget::SceneTo
 {
     /// \todo replace EditMode with suitable subclasses
     tool->addButton (new InstanceMode (this, tool), "object");
-    tool->addButton (
-        new EditMode (this, QIcon (":placeholder"), Mask_Pathgrid, "Pathgrid editing"),
-        "pathgrid");
+    tool->addButton (new PathgridMode (this, tool), "pathgrid");
 }
 
 CSMDoc::Document& CSVRender::WorldspaceWidget::getDocument()
 {
     return mDocument;
 }
 
-osg::Vec3f CSVRender::WorldspaceWidget::getIntersectionPoint (const QPoint& localPos,
-    unsigned int interactionMask, bool ignoreHidden) const
+CSVRender::WorldspaceHitResult CSVRender::WorldspaceWidget::mousePick (const QPoint& localPos,
+    unsigned int interactionMask) const
 {
     // (0,0) is considered the lower left corner of an OpenGL window
     int x = localPos.x();
     int y = height() - localPos.y();
 
-    osg::ref_ptr<osgUtil::LineSegmentIntersector> intersector (
-        new osgUtil::LineSegmentIntersector (osgUtil::Intersector::WINDOW, x, y));
+    // Get intersection
+    osg::ref_ptr<osgUtil::LineSegmentIntersector> intersector (new osgUtil::LineSegmentIntersector(
+        osgUtil::Intersector::WINDOW, x, y));
 
-    intersector->setIntersectionLimit (osgUtil::LineSegmentIntersector::NO_LIMIT);
-    osgUtil::IntersectionVisitor visitor (intersector);
-
-    unsigned int mask = interactionMask;
-
-    if (ignoreHidden)
-        mask &= getVisibilityMask();
+    intersector->setIntersectionLimit(osgUtil::LineSegmentIntersector::NO_LIMIT);
+    osgUtil::IntersectionVisitor visitor(intersector);
 
-    visitor.setTraversalMask (mask);
+    visitor.setTraversalMask(interactionMask);
 
-    mView->getCamera()->accept (visitor);
+    mView->getCamera()->accept(visitor);
 
-    for (osgUtil::LineSegmentIntersector::Intersections::iterator iter = intersector->getIntersections().begin();
-         iter!=intersector->getIntersections().end(); ++iter)
+    // Get relevant data
+    for (osgUtil::LineSegmentIntersector::Intersections::iterator it = intersector->getIntersections().begin();
+         it != intersector->getIntersections().end(); ++it)
     {
+        osgUtil::LineSegmentIntersector::Intersection intersection = *it;
+
         // reject back-facing polygons
```

Comment:
Actually, it seems that occasionally (very rare for me), clicks on the terrain sometimes don't register. I suspect it is caused by the "reject of back-facing polygons" bit. Either that or osg isn't properly raytracing.